### PR TITLE
Upgrade fortran packaging

### PR DESCRIPTION
**Pull request by emmaai**
As the title, use more modern approach to wrap and compile fortran code, also
- remove python 3.8, add python 3.10
- update precommit and GitHub Actions

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -39,18 +39,18 @@ jobs:
       - name: Install conda dependencies
         run: |
           mamba install datacube gdal pycodestyle pylint pytest-cov pytest numexpr lxml pydash python-rapidjson \
-          lxml pydash python-rapidjson ruamel.yaml structlog ciso8601 cattrs boltons dawg
+          lxml pydash python-rapidjson ruamel.yaml structlog ciso8601 cattrs boltons build
           # gdal is required by digitalearthau??
       - name: Install pip deps
         run: |
-          pip install --extra-index-url=https://packages.dea.ga.gov.au/ eodatasets3 digitalearthau
+          pip install --extra-index-url=https://packages.dea.ga.gov.au/ eodatasets3
```
Comment (omad) [inline .github/workflows/tests.yaml:46]: Probably don't want to be pulling from packages.dea.ga.gov.au for eodatasets3. I think it's in PyPI and should be grabbed from there.

**Comment by omad** [comment]
This looks really important Emma. Can you explain some more about what you've implemented and why you've done it this way. (Curiosity not criticism, I'm not at all an expert on these build systems).

From [Status of numpy.distutils and migration advice — NumPy v2.1 Manual](https://numpy.org/doc/stable/reference/distutils_status_migration.html#), I see that:
> [numpy.distutils](https://numpy.org/doc/stable/reference/distutils.html#module-numpy.distutils) has been deprecated in NumPy 1.23.0. It will be removed for Python 3.12;

NumPy 1.23 came out in June 2022.

This FC was being built with `numpy.distutils`, so getting rid of that makes total sense. 

It looks like this PR switches to using [scikit-build-core](https://scikit-build-core.readthedocs.io/en/latest/) + CMake. Why did you pick this combination over meson?

It looks like you've also switched from using [F2Py ](https://numpy.org/doc/stable/f2py/index.html )to a hand written C wrapper of the Fortran functions.

As far as I know, F2Py is still supported/recommended, so what's the reasoning for this change?

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (Ariana-B) [reply .github/workflows/tests.yaml:46]: It could even be grabbed from conda-forge alongside `datacube`!

File: fc/fc_app.py [hunk 2]
```diff
@@ -9,13 +9,12 @@
 4. datacube-fc generate
```
Comment (Ariana-B) [inline fc/fc_app.py:9]: This needs to be updated given that it seems you've removed the `ensure-products` and `generate` cli commands (and also anywhere else that my mention/use them)

**Comment by emmaai** [comment]
re the toolset and approach choice:

- we agreed that `numpy.distutils` has to go
- more general toolset, such as `meson` and `ninja` vs `scikit-setup` and `cmake`, here are how I made my choice
    - `meson` and `ninja` is very good and powerful to compile large and complex projects, though it requires a pre-made building environment. 
    - `scikit-setup` and `cmake` is more aligned with modern python packaging approach, all dependencies are pulled and built in a virtual env invoked by `python -m build` based on the information in `pyproject.toml`, then both libs and packages are build in one go in this env, then the virtual env can be discarded.
    - `scikit-setup` and `cmake` enables naive `pip install .` from the folder without building the package separately with `meson` or `ninja`, it's more convenient in development stage
- `f2py` vs manual c wrapper, got admitted that I failed to call `f2py` in `cmake`, it gave me long list of dependency errors, which was compiler related. I just anticipated that it could happen if the env changed or compiler changed, e.g., the compiler distributed with ubuntu repo vs the one pulled from conda. The support of python extension on fortran is largely up to `numpy.distutils` and honestly I don't know how well/bad it is done with `setuptools`. I didn't dig deeper to figure out why or how to deal with it. As the routine of fortran code is really simple, just one function to call, it looked to me much cheaper of writing a c wrapper than alternatives.

**Review by omad — APPROVED** [review]
Thanks for the explanation Emma!!

Sounds good to me, tested building locally.

**Pull request merged**